### PR TITLE
ES-574 sqlx db

### PR DESCRIPTION
**Pull request by markya0616**
Replace Gorm with sqlx for context support

### CODE REVIEW

File: store/account/account_test.go [hunk 1]
```diff
@@ -62,48 +65,49 @@ var _ = Describe("Account Database Test", func() {
...
-		db.Delete(&model.Account{
-			ContractAddress: model.ETHBytes,
-		})
+		_, err = db.Exec("DELETE FROM erc20")
+		Expect(err).Should(Succeed())
+		// Remove ETH accounts & balances
+		_, err = db.Exec("DELETE FROM accounts")
+		Expect(err).Should(Succeed())
+		_, err = db.Exec("DELETE FROM eth_transfer")
+		Expect(err).Should(Succeed())
```
Comment (kimmylin) [inline store/account/account_test.go]: why didn't we have to delete any erc20 table?

File: store/store.go [hunk 2]
```diff
@@ -66,14 +68,14 @@ type manager struct {
 	headerStore
 	accountStore
 
-	db          *gorm.DB
+	db          DbOrTx
```
Comment (kimmylin) [inline store/store.go]: keep it simple first and use sqlx.DB?

File: store/transaction_receipt/transaction_receipt.go [hunk 3]
```diff
@@ -17,68 +17,148 @@
...
+	Delete(ctx context.Context, from, to int64) (err error)
+	FindReceipt(ctx context.Context, hash []byte) (result *model.Receipt, err error)
 }
 
+const (
+	insertReceiptSQL = "INSERT INTO transaction_receipts (root, status, cumulative_gas_used, bloom, tx_hash, contract_address, gas_used, block_number) VALUES (X'%s', %d, %d, X'%s', X'%s', X'%s', %d, %d)"
+	insertLogSQL     = "INSERT INTO receipt_logs (tx_hash, block_number, contract_address, event_name, topic1, topic2, topic3, data) VALUES (X'%s', %d, X'%s', X'%s', X'%s', X'%s', X'%s', X'%s')"
+	deleteReceiptSQL = "DELETE FROM transaction_receipts WHERE block_number >= %d AND block_number <= %d"
+	deleteLogSQL     = "DELETE FROM receipt_logs WHERE block_number >= %d AND block_number <= %d"
+	findReceiptSQL   = "SELECT * FROM transaction_receipts WHERE `tx_hash` = X'%s'"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/transaction_receipt/transaction_receipt.go]: OK. Update ALL!

File: store/transaction/transaction.go [hunk 4]
```diff
@@ -17,44 +17,62 @@
...
+	Delete(ctx context.Context, from, to int64) (err error)
+	FindTransaction(ctx context.Context, hash []byte) (result *model.Transaction, err error)
+	FindTransactionsByBlockHash(ctx context.Context, blockHash []byte) (result []*model.Transaction, err error)
 }
 
+const (
+	insertSQL                     = "INSERT INTO transactions (`hash`, block_hash, `from`, `to`, nonce, gas_price, gas_limit, amount, payload, block_number) VALUES (X'%s', X'%s', X'%s', X'%s', %d, %d, %d, '%s', X'%s', %d)"
+	deleteSQL                     = "DELETE FROM transactions WHERE block_number >= %d AND block_number <= %d"
+	findTransactionSQL            = "SELECT * FROM transactions WHERE `hash` = X'%s'"
+	findTransactionByBlockHashSQL = "SELECT * FROM transactions WHERE `block_hash` = X'%s'"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/transaction/transaction.go]: OK

File: store/transaction_receipt/transaction_receipt.go [hunk 5]
```diff
@@ -17,68 +17,148 @@
...
+			if err != nil {
+				dbTx.Rollback()
+				return
+			}
+			err = dbTx.Commit()
+		}()
+	} else {
+		dbTx, ok = r.db.(*sqlx.Tx)
+		if !ok {
+			return errors.New("not in a transaction")
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/transaction_receipt/transaction_receipt.go]: Why can it be safer?

File: store/transaction_receipt/transaction_receipt.go [hunk 6]
```diff
@@ -17,68 +17,148 @@
...
 	return &store{
 		db: db,
 	}
 }
 
-func (r *store) Insert(data *model.Receipt) error {
-	// TODO: may need db transaction protection,
-	// but mysql doesn't support nested db transaction
+func (r *store) Insert(ctx context.Context, data *model.Receipt) (err error) {
+	// Ensure we are in a db transaction
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/transaction_receipt/transaction_receipt.go]: We cannot do this. If it's not a db tx, we need to defer tx rollback/commit. It's hard to move to a shard lib.

File: store/subscription/subscription.go [hunk 7]
```diff
@@ -31,45 +36,72 @@ const (
...
 	}
 }
 
-func (t *store) BatchInsert(subs []*model.Subscription) (duplicated []common.Address, err error) {
-	dbTx := t.db.Begin()
+func (t *store) BatchInsert(ctx context.Context, subs []*model.Subscription) (duplicated []common.Address, err error) {
+	db, ok := t.db.(*sqlx.DB)
+	if !ok {
+		return nil, errors.New("already in a transaction")
+	}
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription.go]: OK

File: store/transaction_receipt/transaction_receipt.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (markya0616) [reply store/transaction_receipt/transaction_receipt.go]: OK

File: store/subscription/subscription.go [hunk 8]
```diff
@@ -31,45 +36,72 @@ const (
...
 		if err != nil {
-			log.Error("Failed to commit db", "err", err)
+			duplicated = nil
 		}
 	}()
+
+	nowStr := ToTimeStr(time.Now())
 	for _, sub := range subs {
-		createErr := dbTx.Create(sub).Error
+		_, createErr := dbTx.ExecContext(ctx, fmt.Sprintf(insertSQL, sub.BlockNumber, sub.Group, Hex(sub.Address), nowStr, nowStr))
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription.go]: After discussion, it needs to modify the interface. We can do it in the next PR.

File: store/subscription/subscription.go [hunk 9]
```diff
@@ -31,45 +36,72 @@ const (
...
+	findSQL                   = "SELECT * FROM subscriptions WHERE block_number = %d"
+
+	findOldSubscriptionsSQL    = "SELECT * FROM subscriptions WHERE address IN (%s) AND block_number > 0"
+	findByGroupCntSQL          = "SELECT count(*) FROM subscriptions WHERE `group` = %d"
+	findByGroupLmtSQL          = "SELECT * FROM subscriptions WHERE `group` = %d LIMIT %d, %d"
+	listOldSubscriptionsCntSQL = "SELECT count(*) FROM subscriptions WHERE block_number > 0"
+	listOldSubscriptionsLmtSQL = "SELECT * FROM subscriptions WHERE block_number > 0 LIMIT %d, %d"
+
+	insertTotalBalanceSQL = "INSERT INTO total_balances (token, block_number, `group`, balance, tx_fee, miner_reward, uncles_reward) VALUES (X'%s', %d, %d, '%s', '%s', '%s', '%s')"
+	findTotalBalanceSQL   = "SELECT * FROM total_balances WHERE block_number <= %d AND token = X'%s' AND `group` = %d ORDER BY block_number DESC"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription.go]: OK

File: store/block_header/block_header.go [hunk 10]
```diff
@@ -17,27 +17,41 @@
...
 }
 
+const (
+	insertTdSQL          = "INSERT INTO total_difficulty (block, hash, td) VALUES (%d, X'%s', '%s')"
+	insertSQL            = "INSERT INTO block_headers (hash, parent_hash, uncle_hash, coinbase, root, tx_hash, receipt_hash, difficulty, number, gas_limit, gas_used, time, extra_data, mix_digest, nonce, miner_reward, uncles_inclusion_reward, txs_fee, uncle1_reward, uncle1_coinbase, uncle1_hash, uncle2_reward, uncle2_coinbase, uncle2_hash, created_at) VALUES (X'%s', X'%s', X'%s', X'%s', X'%s', X'%s', X'%s', %d, %d, %d, %d, %d, X'%s', X'%s', X'%s', '%s', '%s', '%s', '%s', X'%s', X'%s', '%s', X'%s', X'%s', '%s')"
+	deleteSQL            = "DELETE FROM block_headers WHERE number >= %d AND number <= %d"
+	findTdSQL            = "SELECT * FROM total_difficulty WHERE hash = X'%s'"
+	findBlockByNumberSQL = "SELECT * FROM block_headers WHERE number = %d"
+	findBlockByHashSQL   = "SELECT * FROM block_headers WHERE hash = X'%s'"
+	findLatestBlockSQL   = "SELECT * FROM block_headers ORDER BY number DESC"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/block_header/block_header.go]: OK

File: store/account/account.go [hunk 11]
```diff
@@ -17,114 +17,181 @@
...
+	createERC20TransferSQL         = "CREATE TABLE `%s` (`block_number` bigint(20) DEFAULT NULL,`tx_hash` varbinary(32) DEFAULT NULL, `from` varbinary(20) DEFAULT NULL, `to` varbinary(20) DEFAULT NULL, `value` varchar(32) DEFAULT NULL, INDEX `block_number` (`block_number`), INDEX `tx_hash` (`tx_hash`), INDEX `from` (`from`), INDEX `to` (`to`))"
+	batchUpdateERC20BlockNumberSQL = "UPDATE erc20 SET block_number = %d WHERE address IN (%s)"
+	findERC20SQL                   = "SELECT * FROM erc20 WHERE address = X'%s'"
+	listERC20SQL                   = "SELECT * FROM erc20"
+	listOldERC20SQL                = "SELECT * FROM erc20 WHERE block_number > 0"
+	listNewERC20SQL                = "SELECT * FROM erc20 WHERE block_number = 0"
+	insertAccountSQL               = "INSERT INTO `%s` (block_number, address, balance) VALUES (%d, X'%s', '%s')"
+	findAccountSQL                 = "SELECT * FROM `%s` WHERE address = X'%s' ORDER BY block_number DESC"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/account/account.go]: OK

File: store/account/account.go [hunk 12]
```diff
@@ -17,114 +17,181 @@
...
+	findAccountByNumberSQL         = "SELECT * FROM `%s` WHERE address = X'%s' AND block_number <= %d ORDER BY block_number DESC"
+	deleteAccountsSQL              = "DELETE FROM `%s` WHERE block_number >= %d AND block_number <= %d"
+	insertTransferSQL              = "INSERT INTO `%s` (block_number, tx_hash, `from`, `to`, value) VALUES (%d, X'%s', X'%s', X'%s', '%s')"
+	findTransferSQL                = "SELECT * FROM `%s` WHERE (`from` = X'%s' OR `to` = X'%s') ORDER BY block_number DESC"
+	findTransferByNumberSQL        = "SELECT * FROM `%s` WHERE (`from` = X'%s' OR `to` = X'%s') AND block_number <= %d ORDER BY block_number DESC"
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/account/account.go]: OK

File: store/account/account_test.go (hunk 1 above)
Comment (markya0616) [reply store/account/account_test.go]: We drop them.
https://github.com/getamis/eth-indexer/pull/100/files/68dc4ecdab129115797ce73ffb7093dbe0a7b0a6..c60c147d5edf8262bae0dae91ef3172715ffd9cd#diff-1d2f602ad7e8a4ec7bfadb89f31ba418R87

File: store/store.go (hunk 2 above)
Comment (markya0616) [reply store/store.go]: OK

**Pull request closed without merging**